src/opnmidi.cpp

```diff
@@ -514,24 +521,100 @@ OPNMIDI_EXPORT void opn2_setDebugMessageHook(struct OPN2_MIDIPlayer *device, OPN
 }
 
 
+template <class Dst>
+static void CopySamplesRaw(OPN2_UInt8 *dstLeft, OPN2_UInt8 *dstRight, const int32_t *src,
+                           size_t frameCount, unsigned sampleOffset)
+{
+    for(size_t i = 0; i < frameCount; ++i) {
+        *(Dst *)(dstLeft + i * sampleOffset) = src[2 * i];
+        *(Dst *)(dstRight + i * sampleOffset) = src[2 * i + 1];
+    }
+}
 
-inline static void SendStereoAudio(int      &samples_requested,
-                                   ssize_t  &in_size,
-                                   short    *_in,
-                                   ssize_t   out_pos,
-                                   short    *_out)
+template <class Dst, class Ret>
+static void CopySamplesTransformed(OPN2_UInt8 *dstLeft, OPN2_UInt8 *dstRight, const int32_t *src,
+                                   size_t frameCount, unsigned sampleOffset,
+                                   Ret(&transform)(int32_t))
+{
+    for(size_t i = 0; i < frameCount; ++i) {
+        *(Dst *)(dstLeft + i * sampleOffset) = transform(src[2 * i]);
+        *(Dst *)(dstRight + i * sampleOffset) = transform(src[2 * i + 1]);
+    }
+}
+
+static int SendStereoAudio(int         samples_requested,
+                           ssize_t     in_size,
+                           int32_t    *_in,
+                           ssize_t     out_pos,
+                           OPN2_UInt8 *left,
+                           OPN2_UInt8 *right,
+                           const OPNMIDI_AudioFormat *format)
 {
     if(!in_size)
-        return;
-    size_t offset       = static_cast<size_t>(out_pos);
+        return 0;
+    size_t outputOffset = static_cast<size_t>(out_pos);
     size_t inSamples    = static_cast<size_t>(in_size * 2);
-    size_t maxSamples   = static_cast<size_t>(samples_requested) - offset;
+    size_t maxSamples   = static_cast<size_t>(samples_requested) - outputOffset;
     size_t toCopy       = std::min(maxSamples, inSamples);
-    std::memcpy(_out + out_pos, _in, toCopy * sizeof(short));
+
+    OPNMIDI_SampleType sampleType = format->type;
+    const unsigned containerSize = format->containerSize;
+    const unsigned sampleOffset = format->sampleOffset;
+
```

Comment:
You must to add here next lines:
```cpp
left  += outputOffset * containerSize;
right += outputOffset * containerSize;
```
or crap will be generated when requested buffer size is bigger than 1024 samples (512 frames).

Reply:
Oooooops. I left a big mistake behind me before going for a nap. damn

Thankfully you do good job of cleaning up behind me.
I couldn't get this problem because I work in small buffer sizes.

Anyway, the step is not **containerSize** it's **sampleOffset**.
I find it bizarre that this would work.
I think it's supposed to be
```
left  += outputOffset * sampleOffset;
right += outputOffset * sampleOffset;
```

In interleaved configuration sampleOffset = 2 * containerSize. It's the amount to skip to go to the channel's next one.
In non-interleaved that would be sampleOffset = 1 * containerSize (ie. what you wrote)

Surprising, because we don't have a use case using non-interleaved. (other than jack code to come in the future because jack is non-interleaved output)

Reply:
To add to the above: actually, this may in fact be

```
left  += (outputOffset / 2) * sampleOffset;
right +=(outputOffset / 2) * sampleOffset;
```

Because outputOffset get its value from gotten_len, a value expressed in samples not frames.

This confusion of samples/frames units in general is terrible. It's why I think audio programming should always speak in terms of frames.